query-engine/core/src/query_graph_builder/extractors/filters/mod.rs

```diff
@@ -144,7 +144,133 @@ pub fn extract_filter(value_map: ParsedInputMap, model: &ModelRef) -> QueryGraph
             _ => Ok(Filter::and(filters)),
         }
     }
-    extract_filter(value_map, model, 0)
+
+    let filter = extract_filter(value_map, model, 0)?;
+    let filter = merge_search_filters(filter);
+
+    Ok(filter)
+}
+
+/// Search filters that have the same query and that are in the same condition block
+/// are merged together to optimize the generated SQL statements.
+/// This is done in three steps (below transformations are using pseudo-code):
+/// 1. We flatten the filter tree.
+/// eg: `Filter(And([ScalarFilter, ScalarFilter], And([ScalarFilter])))` -> `Filter(And([ScalarFilter, ScalarFilter, ScalarFilter]))`
+/// 2. We index search filters by their query.
+/// eg: `Filter(And([SearchFilter("query", [FieldA]), SearchFilter("query", [FieldB])]))` -> `{ "query": [FieldA, FieldB] }`
+/// 3. We reconstruct the filter tree and merge the search filters that have the same query along the way
+/// eg: `Filter(And([SearchFilter("query", [FieldA]), SearchFilter("query", [FieldB])]))` -> `Filter(And([SearchFilter("query", [FieldA, FieldB])]))`
+fn merge_search_filters(filter: Filter) -> Filter {
+    // The filter tree _needs_ to be flattened for the merge to work properly
+    let flattened = flatten_filter(filter);
+
+    match flattened {
+        Filter::And(and) => Filter::And(do_merge_search_filters(and)),
+        Filter::Or(or) => Filter::Or(do_merge_search_filters(or)),
+        Filter::Not(not) => Filter::Not(do_merge_search_filters(not)),
+        _ => flattened,
+    }
+}
+
+fn do_merge_search_filters(filters: Vec<Filter>) -> Vec<Filter> {
+    let mut filters_by_val: HashMap<PrismaValue, &Filter> = HashMap::new();
+    let mut projections_by_val: HashMap<PrismaValue, Vec<ScalarProjection>> = HashMap::new();
+    let mut output: Vec<Filter> = vec![];
+
+    // Gather search filters that have the same condition
+    for filter in filters.iter() {
+        match filter {
+            Filter::Scalar(ref sf) => match sf.condition {
+                ScalarCondition::Search(ref pv, _) => {
+                    // If there's already an entry, then store the "additional" projections that will need to be merged
+                    if let Some(projections) = projections_by_val.get_mut(&pv) {
+                        projections.push(sf.projection.clone());
+                    } else {
+                        // Otherwise, store the first search filter found on which we'll merge the additional projections
+                        projections_by_val.insert(pv.clone(), vec![]);
+                        filters_by_val.insert(pv.clone(), filter);
+                    }
+                }
+                _ => output.push(filter.clone()),
+            },
+            Filter::And(and) => {
+                output.push(Filter::And(do_merge_search_filters(and.clone())));
```

Comment:
You probably thought about this, but cloning the entire subfilters for potentially nothing isn't great. If you'd pass down a reference you could only selectively clone, but of course that may not be possible because of the recursion interface.

Reply:
However, if you remember, I also told you that the size of filters in practice (only a few really) doesn't really justify bending over backwards in code to optimize the clones, so it's still fine if this doesn't work.

Reply:
> I also told you that the size of filters in practice (only a few really) doesn't really justify bending over backwards in code to optimize the clones, so it's still fine if this doesn't work

Yeah, that's why I initially removed all references, but given how I refactored the code, passing the vec by value was stupid. It's now fixed, thank you for the suggestion.